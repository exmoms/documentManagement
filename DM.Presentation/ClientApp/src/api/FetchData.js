/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
