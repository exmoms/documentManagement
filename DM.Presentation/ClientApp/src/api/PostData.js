/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
