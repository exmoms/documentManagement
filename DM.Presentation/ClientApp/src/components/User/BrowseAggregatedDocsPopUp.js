/**
 * @file Contains BrowseAggregatedDocsPopUp component implementaion
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @version 1.0
 */

import React, { Component } from "react";
import DocumentTable from "./DocumentTable";
import SearchDocuments from "./SearchDocuments";
import CREATE_DOC_STYLES from "./Styles";
import { withStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogTitle from "@material-ui/core/DialogTitle";
import DialogContent from "@material-ui/core/DialogContent";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContentText from "@material-ui/core/DialogContentText";

import TableContainer from "@material-ui/core/TableContainer";
import Table from "@material-ui/core/Table";
import TableRow from "@material-ui/core/TableRow";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import Paper from "@material-ui/core/Paper";

import { withTranslation } from "react-i18next";

const styles = CREATE_DOC_STYLES;

class BrowseAggregatedDocsPopUp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      obj: {
        id: -1,
        name: "...",
        latest_version: -1,
      },
      data: [],
      columns: [],
    };
    this.updateParentState = this.props.action;
  }
  t = this.props["t"];

  getDocumentList = (list, columns) => {
    this.setState({
      data: list,
      columns: columns,
    });
  };

  toggleShow = (show) => {
    this.updateParentState("show", show);
  };

  handleSelect = (value) => {
    if (value !== null) {
      this.updateParentState("addDoc", {
        modelId: this.props.modelId,
        selectedDoc: value,
      });

      this.toggleShow(false);
    }
  };

  /**
   * @todo Could we reduce rendering this component, once it renders two times
   * each time we update state
   */
  render() {
    const { classes } = this.props;
    return (
      <div>
        <Dialog
          position="centerCenter"
          open={this.props.show}
          onClose={() => this.toggleShow(false)}
          aria-labelledby="form-dialog-title"
        >
          <DialogTitle id="form-dialog-title">
            {this.t("addaggregateddoc")}
          </DialogTitle>
          <DialogContent>
            <DialogContentText>
              {this.t("dialog_selectaggregateddocument")} <strong>Hint:</strong>{" "}
              {this.t("dialog_previewchooseddocument")}
            </DialogContentText>
            <TableContainer component={Paper}>
              <Table className={classes.table} aria-label="simple table">
                <TableBody>
                  <TableRow key="search-form">
                    <TableCell>
                      <SearchDocuments
                        getDocumentList={this.getDocumentList}
                        action="update-document"
                        modelId={this.props.modelId}
                      />
                      <DocumentTable
                        data={this.state.data}
                        columns={this.state.columns}
                        action="update-document"
                        handleSelect={this.handleSelect}
                      />
                    </TableCell>
                  </TableRow>
                </TableBody>
              </Table>
            </TableContainer>
          </DialogContent>
          <DialogActions>
            <Button
              id="CancelButton"
              onClick={() => this.toggleShow(false)}
              color="primary"
            >
              {this.t("button_cancel")}
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    );
  }
}

export default withStyles(styles)(withTranslation()(BrowseAggregatedDocsPopUp));
