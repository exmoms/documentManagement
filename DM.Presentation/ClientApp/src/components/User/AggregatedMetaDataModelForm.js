/**
 * @file Contains AggregatedMetaDataModel component implementaion
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @version 1.0
 */

import React, { Component } from "react";
import BrowseAggregatedDocsPopUp from "./BrowseAggregatedDocsPopUp";
import styles from "./Styles";
import { withStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import TableContainer from "@material-ui/core/TableContainer";
import Table from "@material-ui/core/Table";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import Paper from "@material-ui/core/Paper";
import { withTranslation } from "react-i18next";

class AggregatedMetaDataModelForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      numberOfAggregatedModels: this.props.aggregatedDocsInfo.length,
      /**
       * Array of IDs of child documents that want to aggregate
       * list of objects
       * [{
       *   modelId: [int] id of aggregatedModels,
       *   modelName: [string],
       *   childMetaDataModelId: [int]
       *   selectedDOC: [object] {id: [int], name: [string]}
       * }]
       */
      // current avialable document to browse [temp]
      currentModelId: -1,
      showAggregateDocPopUp: false,
    };
    this.updateParentAggregatedDocs = this.props.updater;
  }
  t = this.props["t"];

  componentDidUpdate(prevProps) {
    if (this.props.aggregatedDocsInfo !== prevProps.aggregatedDocsInfo) {
      var temp = this.props.aggregatedDocsInfo.map((aggregatedModel) => {
        return {
          modelId: aggregatedModel.id,
          selectedId: -1,
        };
      });
      this.setState({
        aggregatedDocuments: temp,
        numberOfAggregatedModels: this.props.aggregatedDocsInfo.length,
        // aggregatedModels: this.props.aggregatedDocsInfo
      });
    }
  }

  addAggreDocOnClickHandler = (aggregatedDoc) => {
    this.setState({
      currentModelId: aggregatedDoc.childMetaDataModelId,
      showAggregateDocPopUp: true,
    });
  };

  updateStateFromChild = (attribName, newState) => {
    if (attribName === "show") {
      this.setState({ showAggregateDocPopUp: newState });
    }
    if (attribName === "addDoc") {
      var index = -1;
      index = this.props.aggregatedDocsInfo.findIndex((item) => {
        return (
          item.childMetaDataModelId === newState.modelId &&
          newState.selectedDoc.id !== -1
        );
      });
      if (index === -1) {
        if (process.env.NODE_ENV === "development") {
          console.log(
            "[ERROR] Can't find the corresponding model with ID:",
            newState.modelId
          );
        }
      } else {
        if (process.env.NODE_ENV === "development") {
          console.log(
            "[SUCCESS] Update with the corresponding model with ID",
            newState.modelId,
            "with selected doc id:",
            newState.selectedDoc.id
          );
        }
        var temp = this.props.aggregatedDocsInfo;
        temp[index].selectedDoc = newState.selectedDoc;
        this.updateParentAggregatedDocs(temp);
      }
    }
  };

  renderAggregatedModelsTable = () => {
    const { classes } = this.props;
    return (
      <TableContainer component={Paper}>
        <Table className={classes.table} aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>{this.t("namelabel")} </TableCell>

              <TableCell>{this.t("actions")} </TableCell>

              <TableCell>{this.t("selecteddocumentname")}</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {this.props.aggregatedDocsInfo.map((aggregatedDoc) => (
              <TableRow key={aggregatedDoc.modelId}>
                <TableCell>{aggregatedDoc.modelName}</TableCell>
                <TableCell>
                  <Button
                    id={"addAggregated" + aggregatedDoc.modelId}
                    onClick={() => {
                      this.addAggreDocOnClickHandler(aggregatedDoc);
                    }}
                    variant="outlined"
                    color="primary"
                    size="small"
                    className={classes.buttonTable}
                  >
                    {this.t("addaggregateddoc")}
                  </Button>
                </TableCell>
                <TableCell
                  id={"table-cell-aggregated-doc" + aggregatedDoc.modelId}
                >
                  {aggregatedDoc.selectedDoc.name}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    );
  };

  render() {
    // Check if there is any data to render
    const renderContents = () => {
      if (this.state.numberOfAggregatedModels !== 0) {
        return (
          <div>
            {this.renderAggregatedModelsTable()}
            <BrowseAggregatedDocsPopUp
              modelId={this.state.currentModelId}
              show={this.state.showAggregateDocPopUp}
              action={this.updateStateFromChild}
            />
          </div>
        );
      }
    };
    return <div>{renderContents()}</div>;
  }
}

export default withStyles(styles)(
  withTranslation()(AggregatedMetaDataModelForm)
);
