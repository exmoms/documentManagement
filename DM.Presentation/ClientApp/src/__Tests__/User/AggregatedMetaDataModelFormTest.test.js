import React from "react";
import ReactDom from "react-dom";
import AggregatedMetaDataModelForm from "../../components/User/AggregatedMetaDataModelForm";
import BrowseAggregatedDocsPopUp from "../../components/User/BrowseAggregatedDocsPopUp";
import { getAvailableAggregatedDocs } from "../../api/FetchData";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import Dialog from "@material-ui/core/Dialog";
import DialogContent from "@material-ui/core/DialogContent";
import Autocomplete from "@material-ui/lab/Autocomplete";
afterEach(cleanup);
describe("AggregatedMetaDataModelForm", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    let aggregatedDocsInfo = [
      { modelId: 1, modelName: "doc1", selectedDoc: { id: 2, name: "doc2" } },
    ];
    const updater = jest.fn();
    const div = document.createElement("div");
    ReactDom.render(
      <AggregatedMetaDataModelForm
        updater={updater}
        aggregatedDocsInfo={aggregatedDocsInfo}
      />,
      div
    );
  });

  test("AggregatedMetaDataModelForm matches snapshot", () => {
    let aggregatedDocsInfo = [
      { modelId: 1, modelName: "doc1", selectedDoc: { id: 2, name: "doc2" } },
    ];
    const updater = jest.fn();
    const component = renderer.create(
      <AggregatedMetaDataModelForm
        updater={updater}
        aggregatedDocsInfo={aggregatedDocsInfo}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("AggregatedMetaDataModelForm addAggreDocOnClickHandler functionality", () => {
    let aggregatedDocsInfo = [
      {
        modelId: 1,
        modelName: "doc1",
        selectedDoc: { id: 2, name: "doc2" },
        childMetaDataModelId: 1,
      },
    ];
    const updater = jest.fn();
    let wrapper = mount(
      <AggregatedMetaDataModelForm
        updater={updater}
        aggregatedDocsInfo={aggregatedDocsInfo}
      />
    );
    //test the intialization
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().modelId).toBe(-1);
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().show).toBeFalsy();
    //test addAggreDocOnClickHandler
    act(() => {
      wrapper
        .find("#addAggregated1")
        .at(0)
        .props()
        .onClick({ target: { value: "" } });
    });
    wrapper.update();
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().modelId).toBe(1);
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().show).toBeTruthy();
  });

  test("AggregatedMetaDataModelForm updateStateFromChild functionality when attribName == show", () => {
    let aggregatedDocsInfo = [
      {
        modelId: 1,
        modelName: "doc1",
        selectedDoc: { id: 2, name: "doc2" },
        childMetaDataModelId: 1,
      },
    ];
    const updater = jest.fn();
    let wrapper = mount(
      <AggregatedMetaDataModelForm
        updater={updater}
        aggregatedDocsInfo={aggregatedDocsInfo}
      />
    );
    //test the intialization
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().modelId).toBe(-1);
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().show).toBeFalsy();
    //click updateStateFromChild
    let attribName = "show";
    let newState = true;
    act(() => {
      wrapper
        .find(BrowseAggregatedDocsPopUp)
        .at(0)
        .props()
        .action(attribName, newState);
    });
    wrapper.update();
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().show).toBeTruthy();
  });

  test("AggregatedMetaDataModelForm updateStateFromChild functionality when attribName == addDoc", () => {
    let aggregatedDocsInfo = [
      {
        modelId: 1,
        modelName: "doc1",
        selectedDoc: { id: 2, name: "doc2" },
        childMetaDataModelId: 1,
      },
    ];
    const updater = jest.fn();
    const mockConsole = jest.spyOn(console, "log");
    let wrapper = mount(
      <AggregatedMetaDataModelForm
        updater={updater}
        aggregatedDocsInfo={aggregatedDocsInfo}
      />,
      { attachTo: document.body }
    );
    //test the intialization
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().modelId).toBe(-1);
    expect(wrapper.find(BrowseAggregatedDocsPopUp).props().show).toBeFalsy();
    //click updateStateFromChild
    let attribName = "addDoc";
    let newState = {
      modelId: 1,
      modelName: "doc1",
      selectedDoc: { id: 2, name: "AddedDoc" },
      childMetaDataModelId: 1,
      selectedId: 2,
    };
    act(() => {
      wrapper
        .find(BrowseAggregatedDocsPopUp)
        .at(0)
        .props()
        .action(attribName, newState);
    });
    wrapper.update();
    //commented by the developer
    // expect(document.getElementById("table-cell-aggregated-doc" + newState.modelId).innerHTML).toBe(newState.selectedDoc.name);
    var temp = aggregatedDocsInfo;
    var index = -1;
    index = aggregatedDocsInfo.findIndex((item) => {
      return item.modelId === newState.modelId && newState.selectedId !== -1;
    });
    temp[index].selectedDoc = newState.selectedDoc;
    expect(updater).toHaveBeenCalledTimes(1);
    expect(updater.mock.calls[0][0]).toBe(temp);

    // test console.log statement
    expect(mockConsole).toHaveBeenCalledTimes(0);
  });
});
