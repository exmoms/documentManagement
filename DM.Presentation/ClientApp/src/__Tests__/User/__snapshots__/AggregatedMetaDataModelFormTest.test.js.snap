// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregatedMetaDataModelForm AggregatedMetaDataModelForm matches snapshot 1`] = `
<div>
  <div>
    <div
      className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        aria-label="simple table"
        className="MuiTable-root withI18nextTranslation(AggregatedMetaDataModelForm)-table-6"
      >
        <thead
          className="MuiTableHead-root"
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              namelabel
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              actions
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              selecteddocumentname
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
        >
          <tr
            className="MuiTableRow-root"
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              doc1
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-outlined withI18nextTranslation(AggregatedMetaDataModelForm)-buttonTable-11 MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                disabled={false}
                id="addAggregated1"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  addaggregateddoc
                </span>
              </button>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
              id="table-cell-aggregated-doc1"
            >
              doc2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div />
  </div>
</div>
`;
